core/Enclave/enclave_hsm.cpp

```diff
@@ -458,27 +540,111 @@ sgx_status_t enclave_rsa_encrypt(const uint8_t *cmk_blob,
 {
-    const rsa_key_data_t *temp_pub_key = (rsa_key_data_t *)cmk_blob;
-
-    BIO *bio = BIO_new_mem_buf(temp_pub_key->rsa_public_key, -1);
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
 
-    RSA* rsa = PEM_read_bio_RSAPublicKey(bio, &rsa, NULL, g_domain_key);
+    if (keyspec == NULL || padding_mode == NULL)
+        return SGX_ERROR_INVALID_PARAMETER;
 
-    if(rsa == NULL)
+    uint32_t sealed_rsa_len;
+    switch(keyspec)
     {
-        printf("failed to load rsa key\n");
+        case EH_RSA_2048:
+            sealed_rsa_len = sgx_calc_gcm_data_size(0, 2105);
+
+            if (plaintext_len > RSA_OAEP_2048_SHA_256_MAX_ENCRYPTION_SIZE) {
+                printf("ecall rsa_encrypt plain len is up to 190B.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+
+            if (ciphertext == NULL || ciphertext_len < RSA_OAEP_2048_CIPHER_LENGTH) {
+                printf("ecall rsa_encrypt ciphertext len is too small.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+            break;
+        case EH_RSA_3072:
+            sealed_rsa_len = sgx_calc_gcm_data_size(0, 3060);
+
+            if (plaintext_len > RSA_OAEP_3072_SHA_256_MAX_ENCRYPTION_SIZE) {
+                printf("ecall rsa_encrypt plain len is up to 318B.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+
+            if (ciphertext == NULL || ciphertext_len < RSA_OAEP_3072_CIPHER_LENGTH) {
+                printf("ecall rsa_encrypt ciphertext len is too small.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+            break;
+    }
+    
+    if (UINT32_MAX == sealed_rsa_len) {
+        printf("ecall rsa_encrypt failed to calculate sealed data size.\n");
+        return SGX_ERROR_UNEXPECTED;
+    }
+
+    if (cmk_blob == NULL || cmk_blob_size < sealed_rsa_len) {
+        printf("ecall rsa_encrypt cmk_blob_size is too small.\n");
         return SGX_ERROR_INVALID_PARAMETER;
     }
 
-    switch (keyspec)
-    {
-        case RSA_NO_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_NO_PADDING);
-        case RSA_PKCS1_OAEP_PADDING:
-            RSA_public_encrypt(plaintext_len, plaintext, ciphertext, rsa, RSA_PKCS1_OAEP_PADDING);
+    if (plaintext == NULL || plaintext_len == 0) {
+        printf("ecall rsa_encrypt plaintext or len is wrong.\n");
+        return SGX_ERROR_INVALID_PARAMETER;
     }
 
-    return SGX_SUCCESS;
+    // encrypt
+    char* enc_key = NULL;
+    char* rsa_public_key = NULL;
+    BIO *bio = NULL;
+    RSA* rsa = NULL;
+    uint32_t public_key_size;
+    
+    do {
+        // load rsa public key
+        switch(keyspec)
+        {
+            case EH_RSA_2048:
+                public_key_size = 426;
+                break;
+            case EH_RSA_3072:
+                public_key_size = 601;
+                break;
+        }
+        enc_key = (char*)malloc(cmk_blob_size);
+        sgx_gcm_decrypt(&g_domain_key,
```

Comment:
sgx_gcm_decrypt is a wrapper for the sdk API implement by us, right?
if so, please rename it, don't use "sgx_xxx"

Reply:
check return value

Reply:
> check return value

yes, I'll add

> sgx_gcm_decrypt is a wrapper for the sdk API implement by us, right? if so, please rename it, don't use "sgx_xxx"

the function name will change.